contributors/design-proposals/kubelet-server-certificate-bootstrap-rotation.md

```diff
@@ -0,0 +1,97 @@
+issuing a new Certificate Signing Request to the API Server, safely updating the
+cert/key pair on disk, begin using the new cert/key pair.
+ 
+1. Store the new cert/key pairs in the directory specified by `--cert-dir`, or the
+   default value if it is unspecified. This will allow the kubelet to have a
+   place for storing the multiple cert/key pairs that it might have available at
+   any given moment (because of a rotation in progress).
+    - When cert/key data is directly embedded in the kubelet config file
+      (client-certificate-data, client-key-data)
+        - The values be used only if there are no updated cert/key pairs due to
+          rotation
+    - When cert/key files are specified in kubeconfig, or with the
+      `--tls-cert-file` and `--tls-private-key-file`:
+        - These values be used only if there are no updated cert/key pairs due
+          to rotation
+1. There will be a feature gate to enable certificate bootstrapping and
+   rotation, `RotateKubeletServerCertificate`
+1. Centralize certificate access within the kubelet code to the
+   CertificateManager. The CertificateManager will be responsible for:
+    - Providing the correct certificate to use for establishing TLS connections.
+    - Generating new private keys and requesting new certificates when the
+      current certificate approaches expiry.
+    - Since certificates can rotate at any time, all other parts of the kubelet
+      should ask the CertificateManager for the correct certificate each time a
+      certificate is used. No certificate caching except by the
+      CertificateManager.
+        - TLS connections should prefer to set the
+          [`GetCertificate`](https://golang.org/pkg/crypto/tls/#Config) and
+          [`GetClientCertificate`](https://golang.org/pkg/crypto/tls/#Config)
+          callbacks so that the connection dynamically requests the certificate
+          as needed.
+    - Recovering from kubelet crashes or restarts that occur while certificate
+      transitions are in flight (request issued, but not yet signed, etc)
+1. The RBAC role for nodes will be updated to allow nodes to request new
+   certificates.
+1. Have the CertificateManager repeat the CSR process as certificate expiration
+   approaches.
+    - New certificates will be requested when the configured duration threshold
+      has been exceeded.
+    - Crash safe file structure
+        - Private key for requests in flight are held in memory. In case of
+          interruption, they will be abandoned.
+        - When the corresponding signed certificate is received, the cert/key
+          pair will be written to a single file, eg.
+          kubelet-server-<timestamp>.pem.
+        - Write kubelet-server-updated.pem symlink to point to the new cert/key
+          pair.
+        - Delete kubelet-server-current.pem
+        - Move kubelet-server-updated.pem to kubelet-server-current.pem
+1. Currently the Certificate Request Signing API is hard coded to issue
```

Comment:
how do you go about loading the new key? Is the application watching the file? Or is there some other part of this where we HUP the process when after the move to kubelet-server-current.pem?